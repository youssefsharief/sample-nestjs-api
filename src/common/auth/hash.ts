import * as bcrypt from 'bcryptjs';
const SALT_WORK_FACTOR = 10;

export const hashModule = {
    isPasswordRight(toBeExaminedPassword: string, realPassword: string) {
        return bcrypt.compare(toBeExaminedPassword, realPassword);
    },

    hash(password: string) {
        return bcrypt.hash(password, SALT_WORK_FACTOR);
    },
};
