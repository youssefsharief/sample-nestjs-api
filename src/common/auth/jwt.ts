import * as jwt from 'jsonwebtoken';
const secret = process.env.secret as string;

export const jwtModule = {
    verify(token: string) {
        return new Promise((resolve, reject) => {
            jwt.verify(token, secret, (err, decoded) => {
                if (err) {
                    return reject(err);
                } else {
                    resolve(decoded);
                }
            });
        });
    },

    sign(_id: string, role: string) {
        return jwt.sign(
            {
                _id,
                role,
            },
            secret,
            {
                expiresIn: 60 * 60 * 24 * 7,
            },
        );
    },
};
