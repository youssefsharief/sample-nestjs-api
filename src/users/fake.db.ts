import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { User } from './schemas/user.schema';

@Injectable()
export class FakeDb {
    constructor(@InjectModel('User') private readonly userModel: Model<User>) {}

    createUser(name: string, email: string, password: string, role: string) {
        return true;
    }
}
