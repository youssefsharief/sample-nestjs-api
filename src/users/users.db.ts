import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { User } from './schemas/user.schema';

@Injectable()
export class UsersDb {
    constructor(@InjectModel('User') private readonly userModel: Model<User>) {}

    createUser(name: string, email: string, password: string, role: string): Promise<User> {
        const newUser = new this.userModel({ name, email, password, role });
        return newUser.save();
    }

    getUserByEmail(email: string) {
        return this.userModel
            .findOne({ email })
            .select('-__v')
            .exec();
    }

    getUserById(_id: string) {
        return this.userModel
            .findOne({ _id })
            .select('-__v')
            .exec();
    }

    updateUserInfo(_id: string, name: string, email: string) {
        return this.userModel
            .findOneAndUpdate({ _id }, { $set: { name, email } }, { new: true })
            .select('-password -__v')
            .exec();
    }

    updateRole(_id: string, role: string) {
        return this.userModel
            .findOneAndUpdate({ _id }, { $set: { role } }, { new: true })
            .select('-password -__v')
            .exec();
    }
}
