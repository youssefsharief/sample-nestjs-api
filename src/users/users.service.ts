import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { SignupDto } from './dto/signup.dto';
import { ROLES } from 'src/common/config/roles-constants';
import { User } from './schemas/user.schema';

@Injectable()
export class UsersService {
    constructor(@InjectModel('User') private readonly userModel: Model<User>) {}

    getHello(): string {
        return 'Hello World!';
    }

    async signup(payload: SignupDto): Promise<User> {
        const createdUser = new this.userModel({ ...payload, role: ROLES.regular });
        return await createdUser.save();
    }

    async getUserByEmail(email: string): Promise<User> {
        return this.userModel
            .findOne({ email })
            .select('-__v')
            .exec();
    }

    async getUserInfo(_id: string): Promise<User> {
        return this.userModel
            .findOne({ _id })
            .select('-__v')
            .exec();
    }

    async updateUserInfo(_id: string, payload: object) {
        return this.userModel.findOneAndUpdate({ _id }, payload, { new: true }).select('-password -__v');
    }
}
