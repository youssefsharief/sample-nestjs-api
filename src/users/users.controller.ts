import {
    Get,
    Controller,
    Post,
    Param,
    Put,
    Body,
    UseGuards,
    UseInterceptors,
    ConflictException,
    UsePipes,
    ValidationPipe,
    UnauthorizedException,
    NotFoundException,
    Patch,
    HttpStatus,
    HttpCode,
} from '@nestjs/common';
import { SignupDto } from './dto/signup.dto';
import { UsersDb } from './users.db';
import { LoginDto } from './dto/login.dto';
import { DurationLoggerInterceptor } from '../common/interceptors/duration-logger.interceptor';
import { clearUnneededDataFromPayload } from './utility/utility-service';
import { UpdateInfoDto } from './dto/update-info.dto';
import { ROLES } from '../common/constants/roles-constants';
import { jwtModule } from '../common/auth/jwt';
import { hashModule } from '../common/auth/hash';
import { AllowValidAuthTokenGuard } from '../common/guards/allow-valid-auth-token.guard';
import { AllowOnlyMyselfGuard } from '../common/guards/allow-only-myself.guard';
import { AllowOnlyManagersGuard } from '../common/guards/allow-only-managers.guard';

@Controller('api/users')
@UseInterceptors(DurationLoggerInterceptor)
export class UsersController {
    constructor(private readonly usersDb: UsersDb) {}

    @UseGuards(AllowValidAuthTokenGuard, AllowOnlyManagersGuard)
    @Get('protected')
    protected(): { protected: string } {
        return { protected: 'yeah' };
    }

    @UseGuards(AllowValidAuthTokenGuard, AllowOnlyManagersGuard)
    @Patch(':id/upgradeRole')
    async upgradeRole(@Param() params: { id: string }) {
        const user = await this.usersDb.updateRole(params.id, ROLES.manager);
        if (!user) {
            throw new NotFoundException('User does not exist');
        }
        return { updated: true };
    }

    @UseGuards(AllowValidAuthTokenGuard, AllowOnlyMyselfGuard)
    @Get(':id/info')
    async getUserInfo(@Param() params: { id: string }) {
        const user = await this.usersDb.getUserById(params.id);
        if (!user) {
            throw new NotFoundException('User does not exist');
        }
        return clearUnneededDataFromPayload(user);
    }

    @UseGuards(AllowValidAuthTokenGuard, AllowOnlyMyselfGuard)
    @Put(':id/info')
    @UsePipes(ValidationPipe)
    async updateUserInfo(@Param() params: { id: string }, @Body() updateInfoDto: UpdateInfoDto) {
        const user = await this.usersDb.updateUserInfo(params.id, updateInfoDto.name, updateInfoDto.email);
        if (!user) {
            throw new NotFoundException('User does not exist');
        } else {
            return { updated: true };
        }
    }

    @Post()
    @UsePipes(ValidationPipe)
    async signup(@Body() signupDto: SignupDto) {
        try {
            const user = await this.usersDb.createUser(
                signupDto.name,
                signupDto.email,
                await hashModule.hash(signupDto.password),
                ROLES.regular,
            );
            return { user: clearUnneededDataFromPayload(user), token: jwtModule.sign(user._id, user.role) };
        } catch (e) {
            if (e.code === 11000) {
                throw new ConflictException('Email already exists');
            } else {
                throw e;
            }
        }
    }

    @Post('login')
    @HttpCode(HttpStatus.OK)
    @UsePipes(ValidationPipe)
    async login(@Body() loginDto: LoginDto) {
        const wrongCredentialsMessage = 'Email or/and password are wrong';
        const user = await this.usersDb.getUserByEmail(loginDto.email);
        if (!user) {
            throw new UnauthorizedException(wrongCredentialsMessage);
        } else {
            const isPasswordRight = await hashModule.isPasswordRight(loginDto.password, user.password);
            if (!isPasswordRight) {
                throw new UnauthorizedException(wrongCredentialsMessage);
            } else {
                return { user: clearUnneededDataFromPayload(user), token: jwtModule.sign(user._id, user.role) };
            }
        }
    }
}
