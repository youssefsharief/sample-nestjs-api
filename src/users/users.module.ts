import { Module } from '@nestjs/common';
import { UsersDb } from './users.db';
import { MongooseModule } from '@nestjs/mongoose';
import { UserSchema } from './schemas/user.schema';
import { UsersController } from './users.controller';

@Module({
    imports: [MongooseModule.forFeature([{ name: 'User', schema: UserSchema }])],
    controllers: [UsersController],
    providers: [UsersDb],
})
export class UsersModule {}
