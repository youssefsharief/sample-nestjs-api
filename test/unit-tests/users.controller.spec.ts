import { Test } from '@nestjs/testing';
import { UsersController } from '../../src/users/users.controller';
import { UsersDb } from '../../src/users/users.db';
import { User } from '../../src/users/schemas/user.schema';
import { hashModule } from '../../src/common/auth/hash';
import { HttpStatus } from '@nestjs/common';

describe('AppController', () => {
    let usersDb: UsersDb;
    let usersController: UsersController;

    beforeEach(async () => {
        const moduleFixture = await Test.createTestingModule({
            controllers: [UsersController],
            providers: [{ provide: UsersDb, useValue: {} }],
        }).compile();

        usersDb = moduleFixture.get(UsersDb);
        usersController = moduleFixture.get(UsersController);
    });

    describe('upgrade role', () => {
        describe('user do not exist', () => {
            it('should return a NotFound Excpetion', async () => {
                try {
                    usersDb.updateRole = jest.fn(() => Promise.resolve(null));
                    await usersController.upgradeRole({ id: '1' });
                } catch (e) {
                    expect(e.message.statusCode).toBe(HttpStatus.NOT_FOUND);
                }
            });
        });

        describe('user exist', () => {
            it('shouldreturn success response', async () => {
                usersDb.updateRole = jest.fn(() =>
                    Promise.resolve({
                        _id: '1',
                        name: 'a cool name',
                        role: 'user',
                        password: 'a password43',
                        email: 'aa@test.com',
                    } as User),
                );
                const updated = await usersController.upgradeRole({ id: '1' });
                expect(updated).toBeTruthy();
            });
        });
    });

    describe('update user info', () => {
        describe('user do not exist', () => {
            it('should return a NotFound Excpetion', async () => {
                try {
                    usersDb.updateUserInfo = jest.fn(() => Promise.resolve(null));
                    await usersController.updateUserInfo({ id: '1' }, { email: 'aa@test.com', name: 'a cool name' });
                } catch (e) {
                    expect(e.message.statusCode).toBe(HttpStatus.NOT_FOUND);
                }
            });
        });

        describe('user exist', () => {
            it('should return token with updated props', async () => {
                usersDb.updateUserInfo = jest.fn(() =>
                    Promise.resolve({
                        _id: '1',
                        name: 'a cool name',
                        role: 'user',
                        password: 'a password43',
                        email: 'aa@test.com',
                    } as User),
                );
                const updated = await usersController.updateUserInfo({ id: '1' }, { email: 'aa@test.com', name: 'a cool name' });
                expect(updated).toBeTruthy();
            });
        });
    });

    describe('signup', () => {
        it('should return token and user object with role, name, and email', async () => {
            usersDb.createUser = jest.fn(() =>
                Promise.resolve({
                    _id: '1',
                    name: 'a cool name',
                    role: 'user',
                    password: 'a password43',
                    email: 'aa@test.com',
                } as User),
            );
            const userDetails = await usersController.signup({ email: 'aa@test.com', password: 'rety', name: 'a cool name' });
            expect(userDetails.token).toBeTruthy();
            expect(userDetails.user).toMatchSnapshot();
        });
    });

    describe('login', () => {
        describe('correct credentials', () => {
            it('should return token and user object with role, name, and email', async () => {
                const hashedPassword = await hashModule.hash('rety');
                usersDb.getUserByEmail = jest.fn(() =>
                    Promise.resolve({
                        _id: '1',
                        name: 'a name',
                        role: 'user',
                        password: hashedPassword,
                        email: 'aa@test.com',
                    } as User),
                );
                const userDetails = await usersController.login({ email: 'aa@test.com', password: 'rety' });
                expect(userDetails.token).toBeTruthy();
                expect(userDetails.user).toMatchSnapshot();
            });
        });

        describe('wrong credentials', () => {
            it('should throw unauthorized exception', async () => {
                usersDb.getUserByEmail = jest.fn(() =>
                    Promise.resolve({
                        _id: '1',
                        name: 'a name',
                        role: 'user',
                        password: 'pass',
                        email: 'aa@test.com',
                    } as User),
                );
                try {
                    await usersController.login({ email: 'aa@test.com', password: 'wrong pass' });
                } catch (e) {
                    expect(e.message.statusCode).toBe(HttpStatus.UNAUTHORIZED);
                    expect(e.message).toMatchSnapshot();
                }
            });
        });

        describe('user does not exist', () => {
            it('should throw unauthorized exception without letting the user know that the user do not exist', async () => {
                usersDb.getUserByEmail = jest.fn(() => Promise.resolve(null));
                try {
                    await usersController.login({ email: 'aa@test.com', password: 'any pass' });
                } catch (e) {
                    expect(e.message.statusCode).toBe(HttpStatus.UNAUTHORIZED);
                    expect(e.message).toMatchSnapshot();
                }
            });
        });
    });

    describe('getUserInfo', () => {
        it('should return userInfo without password', async () => {
            usersDb.getUserById = jest.fn(() =>
                Promise.resolve({ name: 'A clear name', _id: 'as454g', role: 'manager', password: 'a' } as User),
            );
            const actual = await usersController.getUserInfo({ id: 'as454g' });
            expect(actual).toMatchSnapshot();
        });

        describe('user does not exist', () => {
            it('should return 404', async () => {
                usersDb.getUserById = jest.fn(() => Promise.resolve(null));
                try {
                    await usersController.getUserInfo({ id: 'as454g' });
                } catch (e) {
                    expect(e.message.statusCode).toBe(HttpStatus.NOT_FOUND);
                    expect(e.message).toMatchSnapshot();
                }
            });
        });
    });

    describe('protected', () => {
        it('should return', async () => {
            const val = await usersController.protected();
            expect(val).toBeTruthy();
        });
    });
});
