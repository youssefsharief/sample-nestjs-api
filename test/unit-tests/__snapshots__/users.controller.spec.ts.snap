// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController getUserInfo should return userInfo without password 1`] = `
Object {
  "_id": "as454g",
  "name": "A clear name",
  "password": undefined,
  "role": "manager",
}
`;

exports[`AppController getUserInfo user does not exist should return 404 1`] = `
Object {
  "error": "Not Found",
  "message": "User does not exist",
  "statusCode": 404,
}
`;

exports[`AppController login correct credentials should return token and user object with role, name, and email 1`] = `
Object {
  "_id": "1",
  "email": "aa@test.com",
  "name": "a name",
  "password": undefined,
  "role": "user",
}
`;

exports[`AppController login user does not exist should throw unauthorized exception without letting the user know that the user do not exist 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Email or/and password are wrong",
  "statusCode": 401,
}
`;

exports[`AppController login wrong credentials should throw unauthorized exception 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Email or/and password are wrong",
  "statusCode": 401,
}
`;

exports[`AppController signup should return token and user object with role, name, and email 1`] = `
Object {
  "_id": "1",
  "email": "aa@test.com",
  "name": "a cool name",
  "password": undefined,
  "role": "user",
}
`;
